cpp/src/parquet/column_reader.cc

```diff
@@ -1478,16 +1480,28 @@ class TypedRecordReader : public TypedColumnReaderImpl<DType>,
     // We skipped the levels by incrementing 'levels_position_'. For values
     // we do not have a buffer, so we need to read them and throw them away.
     // First we need to figure out how many present/not-null values there are.
-    std::shared_ptr<::arrow::ResizableBuffer> valid_bits;
-    valid_bits = AllocateBuffer(this->pool_);
-    PARQUET_THROW_NOT_OK(valid_bits->Resize(bit_util::BytesForBits(skipped_records),
-                                            /*shrink_to_fit=*/true));
+    int64_t buffer_size = bit_util::BytesForBits(skipped_records);
+    if (valid_bits_for_skip_ == nullptr) {
+      // Preallocate kMaxSkipLevelBufferSize would help minimizing allocations.
+      valid_bits_for_skip_ = AllocateBuffer(
+          this->pool_, std::max<int64_t>(buffer_size, kMaxSkipLevelBufferSize));
+    } else if (buffer_size > kMaxSkipLevelBufferSize) {
+      // Increase the bitmap size.
+      PARQUET_THROW_NOT_OK(valid_bits_for_skip_->Resize(buffer_size,
+                                                        /*shrink_to_fit=*/false));
```

Comment:
A bit weird because it cannot shrink here?

Reply:
the shrink would be meaningless as it's actually expanding?

Reply:
Yeah